aggregator/handler/common/queue.go

```diff
@@ -96,163 +104,163 @@ type forwardHandler struct {
-	if h.closed {
-		h.Unlock()
+func (q *queue) Close() {
+	q.Lock()
+	if q.closed {
+		q.Unlock()
 		return
 	}
-	h.closed = true
-	close(h.bufCh)
-	close(h.closedCh)
-	h.Unlock()
+	close(q.bufCh)
+	close(q.closedCh)
+	q.closed = true
+	q.Unlock()
 
 	// Wait for the queue to be drained.
-	h.wg.Wait()
+	q.wg.Wait()
 }
 
-func (h *forwardHandler) initConnections(servers []string) {
-	h.wg.Add(len(servers))
+func (q *queue) initConnections(servers []string) {
+	q.wg.Add(len(servers))
 	for _, addr := range servers {
-		conn, err := h.tryConnectFn(addr)
+		conn, err := q.tryConnectFn(addr)
 		if err != nil {
-			h.log.WithFields(
+			q.log.WithFields(
 				log.NewField("address", addr),
 				log.NewErrField(err),
 			).Error("error connecting to server")
 		}
-		go h.forwardToConn(addr, conn)
+		go q.forwardToConn(addr, conn)
 	}
 }
 
-func (h *forwardHandler) tryConnect(addr string) (*net.TCPConn, error) {
-	conn, err := h.dialWithTimeoutFn(tcpProtocol, addr, h.connectTimeout)
+func (q *queue) tryConnect(addr string) (*net.TCPConn, error) {
+	conn, err := q.dialWithTimeoutFn(tcpProtocol, addr, q.connectTimeout)
 	if err != nil {
-		h.metrics.connectTimeouts.Inc(1)
+		q.metrics.connectTimeouts.Inc(1)
 		return nil, err
 	}
-	atomic.AddInt32(&h.numOpenConnections, 1)
-	h.metrics.connectSuccesses.Inc(1)
+	atomic.AddInt32(&q.numOpenConnections, 1)
+	q.metrics.connectSuccesses.Inc(1)
 	tcpConn := conn.(*net.TCPConn)
-	tcpConn.SetKeepAlive(h.connectionKeepAlive)
+	tcpConn.SetKeepAlive(q.connectionKeepAlive)
 	return tcpConn, nil
 }
 
-func (h *forwardHandler) enqueue(buffer *aggregator.RefCountedBuffer, dropType dropType) error {
-	h.RLock()
-	if h.closed {
-		h.RUnlock()
+func (q *queue) enqueue(buffer *RefCountedBuffer, dropType dropType) error {
+	q.RLock()
+	if q.closed {
+		q.RUnlock()
 		buffer.DecRef()
-		h.metrics.enqueueErrors.Inc(1)
-		return errHandlerClosed
+		q.metrics.enqueueErrors.Inc(1)
+		return errQueueClosed
 	}
 	for {
 		select {
-		case h.bufCh <- buffer:
-			h.RUnlock()
-			h.metrics.enqueueSuccesses.Inc(1)
+		case q.bufCh <- buffer:
+			q.RUnlock()
+			q.metrics.enqueueSuccesses.Inc(1)
 			return nil
 		default:
 			if dropType == dropCurrent {
-				h.RUnlock()
+				q.RUnlock()
 				buffer.DecRef()
-				h.metrics.enqueueErrors.Inc(1)
-				h.metrics.currentDropped.Inc(1)
-				return errBufferQueueFull
+				q.metrics.enqueueErrors.Inc(1)
+				q.metrics.currentDropped.Inc(1)
+				return errQueueFull
 			}
 		}
 
 		// Drop oldest in queue to make room for new buffer.
 		select {
-		case buf := <-h.bufCh:
+		case buf := <-q.bufCh:
 			buf.DecRef()
-			h.metrics.oldestdDropped.Inc(1)
+			q.metrics.oldestdDropped.Inc(1)
 		default:
 		}
 	}
 }
 
-func (h *forwardHandler) forwardToConn(addr string, conn *net.TCPConn) {
-	defer h.wg.Done()
+func (q *queue) forwardToConn(addr string, conn *net.TCPConn) {
+	defer q.wg.Done()
 
 	var err error
 	continueFn := func(int) bool {
-		h.RLock()
-		closed := h.closed
-		h.RUnlock()
+		q.RLock()
```

Comment:
nit: might be more performant to use an [atomic `bool`](https://godoc.org/go.uber.org/atomic#Bool) here. We can replace the reads of `q.closed` with calls to `Load`, and in `Close` we can use `CAS`.